File: p3/p34337Tests/TestFinal.java
import java.lang.reflect.InvocationTargetException;
import java.util.List;

/**
 * Test multiple servers and patrons
 * 
 * @version 1.0
 */
public class TestFinal {
  /**
   * Mary serves cheese pizza
   * 
   * @param buffet buffet to use
   * 
   * @return runnable thread main
   */
  public static Runnable mary(Buffet buffet) {
    return new Runnable() {
      @Override
      public void run() {
        while (buffet.AddPizza(6, SliceType.Cheese)) {
          Util.randSleep(150);
        }
      }
    };
  }

  /**
   * Nelle serves meat pizza
   * 
   * @param buffet buffet to use
   * 
   * @return runnable thread main
   */
  public static Runnable nelle(Buffet buffet) {
    return new Runnable() {
      @Override
      public void run() {
        while (buffet.AddPizza(8, SliceType.Meat)) {
          Util.randSleep(200);
        }
      }
    };
  }

  /**
   * John serves veggie pizza
   * 
   * @param buffet buffet to use
   * 
   * @return runnable thread main
   */
  public static Runnable john(Buffet buffet) {
    return new Runnable() {
      @Override
      public void run() {
        while (buffet.AddPizza(3, SliceType.Veggie)) {
          Util.randSleep(75);
        }
      }
    };
  }

  /**
   * Genevive serves works pizza
   * 
   * @param buffet buffet to use
   * 
   * @return runnable thread main
   */
  public static Runnable genevive(Buffet buffet) {
    return new Runnable() {
      @Override
      public void run() {
        while (buffet.AddPizza(4, SliceType.Works)) {
          Util.randSleep(100);
        }
      }
    };
  }

  /**
   * Laverne eats any kind of pizza, 2 slices at a time
   * 
   * @param buffet buffet to use
   * 
   * @return runnable thread main
   */
  public static Runnable laverne(Buffet buffet) {
    return new Runnable() {
      @Override
      public void run() {
        List<SliceType> slices;
        while ((slices = buffet.TakeAny(2)) != null) {
          Util.printSlices("Laverne", slices);
          Util.testSlices(slices, 2, false);
          Util.randSleep(50);
        }
      }
    };
  }

  /**
   * Michael eats vegetarian pizza, 3 slices at a time
   * 
   * @param buffet buffet to use
   * 
   * @return runnable thread main
   */
  public static Runnable michael(Buffet buffet) {
    return new Runnable() {
      @Override
      public void run() {
        List<SliceType> slices;
        while ((slices = buffet.TakeVeg(3)) != null) {
          Util.printSlices("Michael", slices);
          Util.testSlices(slices, 3, true);
          Util.randSleep(75);
        }
      }
    };
  }

  /**
   * Kelsey eats vegetarian pizza, 1 slice at a time
   * 
   * @param buffet buffet to use
   * 
   * @return runnable thread main
   */
  public static Runnable kelsey(Buffet buffet) {
    return new Runnable() {
      @Override
      public void run() {
        List<SliceType> slices;
        while ((slices = buffet.TakeVeg(1)) != null) {
          Util.printSlices("Kelsey", slices);
          Util.testSlices(slices, 1, true);
          Util.randSleep(25);
        }
      }
    };
  }

  /**
   * Daren eats any kind of pizza, 2 slices at a time
   * 
   * @param buffet buffet to use
   * 
   * @return runnable thread main
   */
  public static Runnable daren(Buffet buffet) {
    return new Runnable() {
      @Override
      public void run() {
        List<SliceType> slices;
        while ((slices = buffet.TakeAny(2)) != null) {
          Util.printSlices("Daren", slices);
          Util.testSlices(slices, 2, false);
          Util.randSleep(50);
        }
      }
    };
  }

  /**
   * Create, run, and join actor threads
   */
  public static void main(String[] args)
      throws NumberFormatException, NoSuchMethodException, SecurityException, ClassNotFoundException,
      InstantiationException, IllegalAccessException, IllegalArgumentException, InvocationTargetException {
    if (args.length != 1) {
      throw new IllegalArgumentException("Parameter(s): <Buffet Class>");
    }
    // Initialize the buffet our threads are using
    Buffet buffet = Util.getBuffet(args[0], 20);

    // Make a few server threads
    List<Thread> threads = List.of(new Thread(mary(buffet)), new Thread(nelle(buffet)), new Thread(john(buffet)),
        new Thread(genevive(buffet)), new Thread(laverne(buffet)), new Thread(michael(buffet)),
        new Thread(kelsey(buffet)), new Thread(daren(buffet)));
    threads.stream().forEach(Thread::start);

    Util.sleep(10);
    buffet.close();

    // When we're done, join with all the threads
    threads.stream().forEach(t -> {
      try {
        t.join();
      } catch (InterruptedException e) {}});
  }
}

File: p3/p5test-1/testfinal.cpp
/**
 * Multi-threaded program to simulate servers and patrons at a pizza buffet restaurant
 */
#include <iostream>
#include <stdexcept>
#include <cstdlib>
#include <pthread.h>
#include <unistd.h>

using namespace std;

#include "util.h"
#include "buffet.h"

/** Mary serves cheese pizza */
void *mary(void *arg) {
  Buffet *buffet = (Buffet*) arg;
  while (buffet->AddPizza(6, Cheese)) {
    // Rest for a moment, then add some slices
    usleep(rand() % 150);
  }

  return NULL;
}

/** Nelle serves pepperoni pizza */
void *nelle(void *arg) {
  Buffet *buffet = (Buffet*) arg;
  while (buffet->AddPizza(8, Meat)) {
    // Rest for a moment, then add some slices
    usleep(rand() % 200);
  }

  return NULL;
}

/** John serves veggie pizza */
void *john(void *arg) {
  Buffet *buffet = (Buffet*) arg;
  while (buffet->AddPizza(3, Veggie)) {
    // Rest for a moment, then add some slices
    usleep(rand() % 75);
  }

  return NULL;
}

/** Genevive serves works pizza */
void *genevive(void *arg) {
  Buffet *buffet = (Buffet*) arg;
  while (buffet->AddPizza(4, Works)) {
    // Rest for a moment, then add some slices
    usleep(rand() % 100);
  }

  return NULL;
}

/** Laverne eats any kind of pizza, 2 slices at a time */
void *laverne(void *arg) {
  Buffet *buffet = (Buffet*) arg;
  try {
    while (true) {
      vector<SliceType> slices = buffet->TakeAny(2);
      printSlices("Laverne", slices);

      // Rest for a moment, then have some more
      usleep(rand() % 50);
    }
  } catch (const runtime_error& e) {
  }

  return NULL;
}

/** Michal eats vegetarian pizza, 3 slices at a time */
void *michal(void *arg) {
  Buffet *buffet = (Buffet*) arg;
  try {
    while (true) {
      vector <SliceType> slices = buffet->TakeVeg(3);
      printSlices("Michal", slices);

      // Rest for a moment, then have some more
      usleep(rand() % 75);
    }
  } catch (const runtime_error& e) {
  }

  return NULL;
}

/** Kelsey eats vegetarian pizza, 1 slice at a time */
void *kelsey(void *arg) {
  Buffet *buffet = (Buffet*) arg;
  try {
    while (true) {
      vector <SliceType> slices = buffet->TakeVeg(1);
      printSlices("Kelsey", slices);

      // Rest for a moment, then have some more
      usleep(rand() % 25);
    }
  } catch (const runtime_error& e) {
  }

  return NULL;
}

/** Daron eats any kind of pizza, 2 slices at a time */
void *daron(void *arg) {
  Buffet *buffet = (Buffet*) arg;
  try {
    while (true) {
      vector <SliceType> slices = buffet->TakeAny(2);
      printSlices("Daron", slices);

      // Rest for a moment, then have some more
      usleep(rand() % 50);
    }
  } catch (const runtime_error& e) {
  }

  return NULL;
}

int main() {
  // Initialize the monitor our threads are using
  Buffet buffet(20);

  // Make a few agent threads
  const int NOTHREADS = 8;
  pthread_t thread[NOTHREADS];

  if (pthread_create(thread + 0, NULL, mary, &buffet) != 0 ||
      pthread_create(thread + 1, NULL, nelle, &buffet) != 0 ||
      pthread_create(thread + 2, NULL, john, &buffet) != 0 ||
      pthread_create(thread + 3, NULL, genevive, &buffet) != 0 ||
      pthread_create(thread + 4, NULL, laverne, &buffet) != 0 ||
      pthread_create(thread + 5, NULL, michal, &buffet) != 0 ||
      pthread_create(thread + 6, NULL, kelsey, &buffet) != 0 ||
      pthread_create(thread + 7, NULL, daron, &buffet) != 0) {
    fail("Can't make a thread we need");
  }

  // Buffet is open for 10 seconds, eat fast
  sleep(10);
  buffet.close();

  // When we're done, join with all the threads
  for (int i=0; i < NOTHREADS; i++) {
    pthread_join(thread[i], NULL);
  }

  return 0;
}
